clippy_lints/src/loops.rs

```diff
@@ -2367,17 +2367,50 @@ fn check_infinite_loop<'a, 'tcx>(cx: &LateContext<'a, 'tcx>, cond: &'tcx Expr, e
         return;
     };
     let mutable_static_in_cond = var_visitor.def_ids.iter().any(|(_, v)| *v);
-    if no_cond_variable_mutated && !mutable_static_in_cond {
+
+    let mut has_break_or_return_visitor = HasBreakOrReturnVisitor {
+        has_break_or_return: false,
+    };
+    has_break_or_return_visitor.visit_expr(expr);
+    let has_break_or_return = has_break_or_return_visitor.has_break_or_return;
+
+    if no_cond_variable_mutated && !mutable_static_in_cond && !has_break_or_return {
         span_lint(
             cx,
             WHILE_IMMUTABLE_CONDITION,
             cond.span,
-            "Variable in the condition are not mutated in the loop body. \
-             This either leads to an infinite or to a never running loop.",
+            "variables in the condition are not mutated in the loop body. \
+             This may lead to an infinite or to a never running loop",
         );
```

Comment:
I still don't like, that we stop linting when the loop body contains a `return` or `break` statement. I have an idea how to solve this in this PR:

```rust
span_lint_and_then(
    cx,
    WHILE_IMMUTABLE_CONDITION,
    cond.span,
    "variables in the condition are not mutated in the loop body",
    |db| {
        db.note("this may lead to an infinite or to a never running loop");

        if has_break_or_return {
            db.note("this loop contains `return`s or `break`s");
            db.help("rewrite it as `if cond { loop { } }`");
        }
    },
);
```

That way, we get a better error message, but avoid false negatives:
```
error: variable in the condition are not mutated in the loop body
   --> src/stream.rs:717:19
    |
717 |             while should_skip {
    |                   ^^^^^^^^^^^
    |
    = note: this may lead to an infinite or to a never running loop
    = note: this loop contains `return`s or `break`s
    = help: rewrite it as `if cond { loop { } }`
```